src/mame/tandberg/tdv2100_disp_logic.cpp

```diff
@@ -0,0 +1,1063 @@
+		}
+	}
+}
+
+INPUT_CHANGED_MEMBER(tandberg_tdv2100_disp_logic_device::uart_changed)
+{
+	set_uart_state_from_switches();
+}
+
+void tandberg_tdv2100_disp_logic_device::set_uart_state_from_switches()
+{
+	int baud_setting = m_sw_rs232_baud->read();
+	int u46_divisor = (baud_setting == 0x0) ? 15 : 11;
+	int u22_divisor = 256>>baud_setting;
+	m_uart_clock->set_unscaled_clock(DOT_CLOCK_3 / (u46_divisor * u22_divisor));
+
+	int rs232_settings = m_sw_rs232_settings->read();
+	m_uart->write_cs(true);
+	m_uart->write_nb1(BIT(m_dsw_u39->read(), 9));
+	m_uart->write_nb2(true);
+	m_uart->write_np(BIT(rs232_settings, 0));
+	m_uart->write_eps(BIT(rs232_settings, 1));
+	m_uart->write_tsb(BIT(rs232_settings, 2));
+}
+
+void tandberg_tdv2100_disp_logic_device::clock_on_line_flip_flop()
+{
+	bool force_on_line = BIT(m_dsw_u39->read(), 5);
+	if(force_on_line)
+	{
+		m_data_terminal_ready = true;
+	}
+	else
+	{
+		// Try to toggle RS232 DTR
+		if(m_rx_handshake)
+		{
+			m_data_terminal_ready = false;
+		}
+		else
+		{
+			m_data_terminal_ready = !m_data_terminal_ready;
+		}
+	}
+
+	int rs232_settings = m_sw_rs232_settings->read();
+	if(!m_data_terminal_ready)
+	{
+		bool auto_tx_handshake = BIT(rs232_settings, 5);
+		m_request_to_send = false;
+		m_rs232->write_rts((m_request_to_send) ? 1 : 0);
+		if(auto_tx_handshake)
+		{
+			m_tx_handshake = m_request_to_send;
+		}
+	}
+	m_rs232->write_dtr((m_data_terminal_ready) ? 1 : 0);
+
+	bool auto_rx_handshake = BIT(rs232_settings, 4);
+	if(auto_rx_handshake)
+	{
+		m_rx_handshake = m_data_terminal_ready;
+	}
+	bool hold_line_key_for_rts = ((m_dsw_u73->read()&0x3) == 0x1);
+	if(m_data_terminal_ready && hold_line_key_for_rts)
+	{
+		if(!m_trans_key_held)
+		{
+			clock_transmit_flip_flop();
+		}
+	}
+	update_rs232_lamps();
+}
+
+void tandberg_tdv2100_disp_logic_device::clock_transmit_flip_flop()
+{
+	bool force_on_line = BIT(m_dsw_u39->read(), 5);
+	if(force_on_line)
+	{
+		m_request_to_send = true;
+	}
+	else if(!m_data_terminal_ready)
+	{
+		m_request_to_send = false;
+	}
+	else
+	{
+		// Try to toggle RS232 RTS
+		if(m_tx_handshake)
+		{
+			m_request_to_send = false;
+		}
+		else
+		{
+			m_request_to_send = !m_request_to_send;
+		}
+	}
+	m_rs232->write_rts((m_request_to_send) ? 1 : 0);
+
+	bool auto_tx_handshake = BIT(m_sw_rs232_settings->read(), 5);
+	if(auto_tx_handshake)
+	{
+		m_tx_handshake = m_request_to_send;
+	}
+	update_rs232_lamps();
+}
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// Strobes for Keyboard
+//
+
+void tandberg_tdv2100_disp_logic_device::process_keyboard_char(uint8_t key)
+{
+	m_data_kbd = key&0x7f;
+	if(key&0x80)
+	{
+		m_data_pending_kbd = true;
+		bool tty_mode = BIT(m_dsw_u39->read(), 4);
+		if(tty_mode && m_uart->tbmt_r())
+		{
+			uart_tx(1);
+		}
+		else
+		{
+			// TODO: Keyboard interrupt for CPU interface
+		}
+	}
+}
+
+void tandberg_tdv2100_disp_logic_device::break_w(int state)
+{
+	m_break_key_held = state;
+}
```

Comment:
If the break key generates a break condition on the RS232 TxD output, shouldn’t this immediately call `m_rs232->write_txd(0)` if the break input is asserted, and restore the output from the AY51013 if it’s deasserted?

With the way you’ve got it now, no break condition will be generated in response to the break key if the AY51013 doesn’t shift anything out.  Similarly, if the break key is released while a break condition is being generated, it won’t end until the AY51013 shifts something out.

Reply:
Good catch, fixed now.